components/basecomponents/MemoryGame/MemoryGameCore.tsx

```diff
@@ -0,0 +1,304 @@
+        setControlsDisabled(true);
+        // play css animations and sounds
+        const { second: card } = selectedCards;
+        flipCard(card!); // 0.3s
+
+        await playCardPhraseCurrentLang(card!);
+        setScene(Scene.resolveCards);
+      },
+      resolveCards: () => {
+        const { first, second } = selectedCards;
+        const cardsMatch = first?.image === second?.image;
+
+        if (cardsMatch) {
+          console.log('cards match');
+          setScene(Scene.cardsMatch);
+        } else {
+          console.log('cards dont match');
+          setScene(Scene.cardsDontMatch);
+        }
+      },
+      cardsMatch: () => {
+        // disable controls
+        setControlsDisabled(true);
+        // play css animations and sounds
+        // cardsMatch animation 0.7s
+        setTimer(() => cardsMatchSound.play(), 100); // sync to animation
+
+        setTimer(() => {
+          setScene(Scene.cardsMatchReward);
+        }, 700);
+      },
+      cardsMatchReward: async () => {
+        // play css animations and sounds
+        Math.random() > 0.5 && (await playPhraseRandomLang(getRandomElement(phrases.good)));
+        // reset selected cards
+        setSelectedCards({ first: null, second: null });
+        // check win
+        setTimer(() => {
+          if (cards.every((card) => card.flipped)) {
+            console.log('win');
+            setScene(Scene.win);
+          } else {
+            setScene(Scene.game);
+          }
+        }, 700);
+      },
+      cardsDontMatch: async () => {
+        // disable controls
+        setControlsDisabled(true);
+        // play animations and sounds
+        await delay(1000);
+        Math.random() > 0.8 && (await playPhraseRandomLang(getRandomElement(phrases.wrong)));
+
+        // setTimer: show cards for some time to remember then flip back
+        const { first, second } = selectedCards;
+        flipCard(first!);
+        flipCard(second!);
+        setSelectedCards({ first: null, second: null });
+        setScene(Scene.cardsDontMatchFlipBack);
+      },
+      cardsDontMatchFlipBack: () => {
+        // wait for cards flip back
+        setTimer(() => {
+          setScene(Scene.game);
+        }, 300);
+      },
+      win: async () => {
+        // disable controls
+        setControlsDisabled(true);
+        // play css animations and sounds
+        await playPhraseRandomLang(getRandomElement(phrases.win));
+        await delay(200);
+        setScene(Scene.winReward);
+      },
+      winReward: () => {
+        // play css animations and sounds
+        winMusic.play();
+      },
+      goNewGame: () => {
+        clearTimers();
+        // play css animations and sounds
+        playPhraseRandomLang(getRandomElement(phrases.newGame));
+        setTimer(() => {
+          setScene(Scene.begin);
+        }, 500);
+      },
+    };
+    console.log(`scene is: ${scene}`);
+    // run scene actions
+    sceneActions[scene]();
+  }, [scene]);
```

Comment:
Okay, you are missing a lot of stuff in the dependency array here. I think the correct way to do this would be move the logic outside useEffect and into a `changeSceneTo` event handler which would execute the side-effects (playing sounds and stuff), then change the state. But that would be a significant refactoring effort, too risky this close to release. If it works, it works, we can improve it later if we need to change this system in the future :)